### DISCUSSION TITLE
Issues in Redis to valkey upgrade

### DISCUSSION THREAD

**Discussion by royatanu94** [valkey-io/valkey#1219] [Q&A]
Hi All,
Getting issue while upgrading redis to valkey in K8s. But it works fine when we fresh install it.

Getting error:
``
cess: exec: \"valkey-server\": executable file not found in $PATH"
 Warning Failed   19m         kubelet      Error: container create failed: time="2024-10-23T07:10:33Z" level=error msg="runc create failed: unable to start container process: exec: \"valkey-server\": executable file not found in $PATH"
 Warning Failed   19m         kubelet      Error: container create failed: time="2024-10-23T07:10:47Z" level=error msg="runc create failed: unable to start container process: exec: \"valkey-server\": executable file not found in $PATH"
``
Please help me to fix it.

**Comment by royatanu94** [comment]
@Rueian @zuiderkwast
Any help is appreciated.

**Reply by rueian** [reply]
You can either
1. add the directory path where your valkey-server executable is located into the PATH environment variable.
2. or just use the absolute path of your valkey-server executable in the command argument of the k8s container.

**Reply by royatanu94** [reply]
@rueian Need some help here also... 
our application client which uses go-redis to communicate, its hosted in k8s and we use valkey-sentinel cluster.. Client was connected to sentinel  & I forcefully rebooted the worker node of our client, that caused some issue in sentinel. 

I could see below logs in sentinel
`
1:X 23 Oct 2024 04:36:07.216 # Failed to resolve hostname 'sentinel-0.sentinel'
1:X 23 Oct 2024 04:36:07.218 # +tilt #tilt mode entered
1:X 23 Oct 2024 04:36:22.236 # Failed to resolve hostname 'sentinel-0.sentinel'
1:X 23 Oct 2024 04:36:22.239 # Error accepting a client connection: (null) (addr= laddr=[::1]:26379)
1:X 23 Oct 2024 04:36:22.239 # Error accepting a client connection: (null) (addr= laddr=[::1]:26379)
1:X 23 Oct 2024 04:36:22.241 # +tilt #tilt mode entered
1:X 23 Oct 2024 04:36:32.257 # Failed to resolve hostname 'redis-0.redis.redis-ptdev2-ns.svc.cluster.local'
1:X 23 Oct 2024 04:36:37.279 # Error accepting a client connection: (null) (addr= laddr=[::1]:26379)
1:X 23 Oct 2024 04:36:37.279 # Error accepting a client connection: (null) (addr=10.244.17.0:49746 laddr=10.244.4.208:26379)
1:X 23 Oct 2024 04:36:37.279 # Error accepting a client connection: (null) (addr=10.244.17.0:39482 laddr=10.244.4.208:26379) `

Any idea how many connections can be established from sentinel & what will happen if the connection forcefully killed ?

**Reply by rueian** [reply]
I think the error you encountered has nothing to do with the limit of connection number. A connection can be killed at anytime and the system should keep running.

The error you saw probably means you restarted a node having a sentinel server running on it, or there were some discontinuations about your k8s DNS resolver .

**Reply by royatanu94** [reply]
@rueian  This happens every time when I reboot our application worker node, but valkey/sentinel node was always up

**Reply by rueian** [reply]
So probably you were hit by the second case.

**Reply by royatanu94** [reply]
@rueian Any suggestion to fix it ? Are you talking about "kube-dns-*" pod which might be running on the worker node which I rebooted ? I also have doubts like core-DNS or Kube-DNS is doing host dns resolution

**Reply by rueian** [reply]
> Are you talking about "kube-dns-*" pod which might be running on the worker node which I rebooted ?

Yes, that is likely the cause. NodeLocal DNSCache https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/ might be helpful.

**Reply by royatanu94** [reply]
@rueian  the worker node running valley did not have any DNS pod restarts so not getting why its showing sentinel address unable to resolve.

**Reply by royatanu94** [reply]
@rueian From logs its seen that my application replica1 discoveres below sentinels
``
redis: 2024/10/25 06:48:18 sentinel.go:742: sentinel: discovered new sentinel="sentinel-2.sentinel:26379" for master="master"
redis: 2024/10/25 06:48:18 sentinel.go:742: sentinel: discovered new sentinel="sentinel-1.sentinel:26379" for master="master"
``
&
replica2 discoveres
``
redis: 2024/10/25 06:47:55 sentinel.go:742: sentinel: discovered new sentinel="sentinel-2.sentinel:26379" for master="master"
redis: 2024/10/25 06:47:55 sentinel.go:742: sentinel: discovered new sentinel="sentinel-0.sentinel:26379" for master="master"
``
Both replica1 and replica2 running in different worker node. Valkey is runnning 3master and 3 sentinel instances in different node.

So, If i reboot my application replica2 worker node, I saw replica1 unable to reach  sentinel-2 & sentinel-1, once after timeouts then it discoveres new sentinel-0 and with that connection works fine.

Immediately after reboot of application replica2 node I have seen below sentinel-0 logs
``
1:X 25 Oct 2024 06:59:25.167 # Failed to resolve hostname 'sentinel-1.sentinel'
1:X 25 Oct 2024 06:59:25.168 # +tilt #tilt mode entered
1:X 25 Oct 2024 06:59:35.207 # Failed to resolve hostname 'sentinel-2.sentinel'
1:X 25 Oct 2024 06:59:35.207 # +tilt #tilt mode entered
1:X 25 Oct 2024 06:59:45.255 # Failed to resolve hostname `'sentinel-2.sentinel'`

Go-redis log in application replica1 :
``
redis: 2024/10/25 06:59:34 sentinel.go:558: sentinel: GetMasterAddrByName master="master" failed: dial tcp: lookup sentinel-2.sentinel: i/o timeout``

**Reply by rueian** [reply]
Can you try deploying your application and valley sentinel to a well-configured K8s? Such as AWS EKS, Azure AKE, or GCP GKE.
